docs/path-resolution.rst

```diff
@@ -0,0 +1,470 @@
+respectively.
+
+Based on how the processing is performed, we can divide imports into two categories:
+
+- :ref:`Direct imports <direct-imports>`, where you specify the full source unit name.
+- :ref:`Relative imports <relative-imports>`, where you specify a path to be combined with the
+  source unit name of the importing file.
+
+.. index:: ! direct import, import; direct
+.. _direct-imports:
+
+Direct Imports
+--------------
+
+An import that does not start with ``./`` or ``../`` is a *direct import*.
+
+::
+
+    import "/project/lib/util.sol";         // source unit name: /project/lib/util.sol
+    import "lib/util.sol";                  // source unit name: lib/util.sol
+    import "@openzeppelin/address.sol";     // source unit name: @openzeppelin/address.sol
+    import "https://example.com/token.sol"; // source unit name: https://example.com/token.sol
+
+After applying any :ref:`import remappings <import-remapping>` the import path simply becomes the
+source unit name.
+
+.. note::
+
+    A source unit name is just an identifier and even if its value happens to look like a path, it
+    is not subject to the normalization rules you would typically expect from filesystem paths.
+    Any ``/./`` or ``/../`` seguments or sequences of multiple slashes remain a part of it.
+    When the source is provided via Standard JSON interface it is entirely possible to associate
+    different content with source unit names that would refer to the same file on disk.
+
+When the source is not available in the virtual filesystem, the compiler passes the source unit name
+to the import callback.
+The Host Filesystem Loader will attempt to use it as a path and look up the file on disk.
+At this point the platform-specific normalization rules kick in and ``/project/lib/math.sol`` and
+``/project/lib/../lib///math.sol`` may actually result in the same file being loaded.
+Note, however, that the compiler will still see them as separate source units that just happen to
+have identical content.
+
+.. index:: ! relative import, ! import; relative
+.. _relative-imports:
+
+Relative Imports
+----------------
+
+An import starting with ``./`` or ``../`` is a *relative import*.
+Such imports specify a path relative to the source unit name of the importing source unit:
+
+.. code-block:: solidity
+    :caption: /project/lib/math.sol
+
+    import "./util.sol" as util;    // source unit name: /project/lib/util.sol
+    import "../token.sol" as token; // source unit name: /project/token.sol
+
+.. code-block:: solidity
+    :caption: lib/math.sol
+
+    import "./util.sol" as util;    // source unit name: lib/util.sol
+    import "../token.sol" as token; // source unit name: token.sol
+
+.. note::
+
+    Note the difference between a *relative import* (e.g. ``import "./util.sol"``) and a
+    *direct import with a relative path* (e.g. ``import "util.sol"``).
+    It is a relative import only if it starts with ``./`` or ``../``.
+    Otherwise it is a direct import.
```

Comment:
I'm worried that the term "direct import with a relative path" could be confusing. Maybe we could say that unless the import starts with `./` or `../` it is a direct import and is thus absolute and imports do not have to start with `/` to be absolute or something...

Reply:
How about this?
```rst
.. note::

    ``import "./util.sol"`` is a relative import but ``import "util.sol"`` is a direct one.
    Relative imports **must** start with ``./`` or ``../``.
    VFS does not have a root directory and absolute paths do not have to start with a slash.
    While both paths would be considered relative in the host filesystem, ``util.sol`` is actually
    absolute in the VFS.
```

Reply:
("must start" sound like it is an error otherwise)
```
.. note::

    Imports that start with `./` or `../` are relative imports, while all others are direct (absolute) imports.
    Direct imports do to need to start with `/`. This, ``import "./util.sol"`` is a relative import but ``import "util.sol"`` is a direct one.
    While both paths would be considered relative in the host filesystem, ``util.sol`` is actually
    absolute in the VFS.
```
"The VFS does not have a root directory" - of course it has, that's where all the topmost files are. You just don't need to use `/` to reference the root directory.

But yeah, this is probably not too important.

Reply:
I really, really do not want to use the term *absolute import* :)

Point taken about `must`. Replaced with `always`, which I hope does not have this connotation.

> "The VFS does not have a root directory" - of course it has, that's where all the topmost files are. You just don't need to use `/` to reference the root directory.

I think one could also argue that all the topmost items are just floating there without any directory containing them but that would be splitting hairs and a root with an empty name makes sense too so I just removed that part for simplicity:)

So here's another attempt:

```rst
.. note::

    Relative imports **always** start with ``./`` or ``../`` so ``import "util.sol"``, unlike
    ``import "./util.sol"``, is a direct import.
    While both paths would be considered relative in the host filesystem, ``util.sol`` is actually
    absolute in the VFS.
```